train/context_interp/trainer.py

```diff
@@ -0,0 +1,152 @@
+import os.path
+import tensorflow as tf
+from data.interp.interp_data import InterpDataSet
+from context_interp.model import ContextInterp
+from train.trainer import Trainer
+from utils.misc import print_progress_bar
+from utils.flow import get_tf_flow_visualization
+from utils.tf import AdamaxOptimizer, optimistic_restore, Logger
+
+
+class ContextInterpTrainer(Trainer):
+    def __init__(self, model, dataset, session, config, verbose=True):
+        super().__init__(model, dataset, session, config, verbose)
+        assert isinstance(self.model, ContextInterp)
+        assert isinstance(dataset, InterpDataSet)
+
+        self.dataset.load(self.session)
+        self.next_sequence_tensor, self.next_sequence_timing_tensor = self.dataset.get_next_batch()
+
+        self.images_a = self.next_sequence_tensor[:, 0]
+        self.images_b = self.next_sequence_tensor[:, 1]
+        self.images_c = self.next_sequence_tensor[:, 2]
+
+        # Get the train network.
+        model_outputs = self.model.get_forward(self.images_a, self.images_c, 0.5, reuse_variables=tf.AUTO_REUSE)
+        self.images_b_pred, self.warped_a_c, self.warped_c_a, self.flow_a_c, self.flow_c_a = model_outputs
+        if self.config['fine_tune']:
+            self.loss = self.model.get_fine_tuning_loss(self.images_b_pred, self.images_b)
+        else:
+            self.loss = self.model.get_training_loss(self.images_b_pred, self.images_b)
+
+        # Get the optimizer.
+        with tf.variable_scope('train'):
+            self.global_step = tf.Variable(initial_value=0, trainable=False, dtype=tf.int32, name='global_step')
+            self.train_op = AdamaxOptimizer(config['learning_rate'], beta1=0.9, beta2=0.999).minimize(
+                self.loss, global_step=self.global_step
+            )
+
+        # Checkpoint saving.
+        self.saver = tf.train.Saver()
+        self.valid_logger = Logger(os.path.join(self.config['checkpoint_directory'], 'valid'), self.session.graph)
```

Comment:
I am confused. This Logger util looks like it would be used for creating summaries for non-tensorflow frameworks. Why use this if we are using tensorflow? It looks like things are being reevaluated just to create np versions to be converted to tf summaries.

Reply:
That's fine I guess. Though it does appear to be quite a bit of code to solve something that isn't much of a problem anyway.

Reply:
The validation summaries weren't working properly (not displaying enough variety of images, way too many unnecessary saves, loss plot was messed up) so I think it's a necessary fix.